### PR TITLE
Implement custom debug for PathBuildingHop

### PR DESCRIPTION
**Pull request by jurvis**
Re-opened of #1300 
This PR is an attempt at #1233.

The goal here is to clean up the existing output for `PathBuildingHop`, and surface only the [properties we care about](https://github.com/lightningdevkit/rust-lightning/issues/1233#issuecomment-1037281691).

## Demo
This is the new output for PathBuildingHop (ran via `routing::router::tests::simple_route_test`):

```
Found a path back to us from the target with 2 hops contributing up to 100 msat: 
[(PathBuildingHop: node_id: NodeId(02531fe6068134503d2723133227c867ac8fa6c83c537e9a44c3c5bdbdcb1fe337), short_channel_id: 2, fee_msat: 100, path_penalty_msat: 0, path_htlc_minimum_msat: 4294967295, cltv_expiry_delta: 83, [8]), (PathBuildingHop: node_id: NodeId(03462779ad4aad39514614751a71085f2f10e1c7a593e4e030efb5b8721ce55b0b), short_channel_id: 4, fee_msat: 100, path_penalty_msat: 0, path_htlc_minimum_msat: 0, cltv_expiry_delta: 65, [32])]
```

## Some Questions
I am somewhat dissatisfied with the current implementation because the print-out just doesn't seem very ergonomic. Ideally, we'll be able to have the print-out be something closer to what [we have for logging routes](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1519). Where instead of printing out the `Vec`, we'll be able to show:

```
Found a path back to us from the target with 2 hops contributing up to 100 msat: 
  node_id: 02531fe6068134503d2723133227c867ac8fa6c83c537e9a44c3c5bdbdcb1fe337, short_channel_id: 2, fee_msat: 100, path_penalty_msat: 0, path_htlc_minimum_msat: 4294967295, cltv_expiry_delta: 8
  node_id: 03462779ad4aad39514614751a71085f2f10e1c7a593e4e030efb5b8721ce55b0b, short_channel_id: 4, fee_msat: 100, path_penalty_msat: 0, path_htlc_minimum_msat: 0, cltv_expiry_delta: 65
```

There is an opportunity for improvement here by adding a macro to `macro_logger` that does a clean Display of `PaymentPath`, but that will mean having to exposing fields in `PathBuildingHop` and `PaymentPath` to be public, which I am not sure we want.

Any thoughts on this will be appreciated!

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
> I am somewhat dissatisfied with the current implementation because the print-out just doesn't seem very ergonomic. Ideally, we'll be able to have the print-out be something closer to what [we have for logging routes](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1519).

Yea, I don't disagree, though I'm not sure how much its worth going the complexification route here, kinda up to you. This is much better than it was already.

> There is an opportunity for improvement here by adding a macro to macro_logger that does a clean Display of PaymentPath, but that will mean having to exposing fields in PathBuildingHop and PaymentPath to be public, which I am not sure we want.

Yea, not a fan of that, if we want to improve the logging I think we should just add a util function in `router.rs` which does the printing given a list of `PathBuildingHop`s, not do it via `Debug` or whatever.

File: lightning/src/routing/router.rs [hunk 1]
```diff
@@ -463,6 +463,13 @@ struct PathBuildingHop<'a> {
 	value_contribution_msat: u64,
 }
 
+impl<'a> core::fmt::Debug for PathBuildingHop<'a> {
+	fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+        write!(f, "PathBuildingHop: node_id: {:?}, short_channel_id: {}, fee_msat: {}, path_penalty_msat: {}, path_htlc_minimum_msat: {}, cltv_expiry_delta: {}", self.node_id, self.candidate.short_channel_id(), self.fee_msat, self.path_penalty_msat, self.path_htlc_minimum_msat, self.candidate.cltv_expiry_delta())?;
```
Comment (jkczyz) [inline lightning/src/routing/router.rs:468]: You can use `Formatter `methods as shown here: https://doc.rust-lang.org/std/fmt/struct.DebugStruct.html#examples

That way pretty-printing is honored if the caller specifies it.

**Comment by jurvis** [comment]
@jkczyz @TheBlueMatt thanks for taking the time to review! I made some changes based on your feedback. I improved the print-out to be prettier, now it looks like that (thanks for that tip on `debug_struct`, Jeff!):

```
TRACE  [lightning::routing::router : lightning/src/routing/router.rs, 1318] Found a path back to us from the target with 2 hops contributing up to 100 msat: 
 [
    PathBuildingHop {
        node_id: NodeId(02531fe6068134503d2723133227c867ac8fa6c83c537e9a44c3c5bdbdcb1fe337),
        short_channel_id: 2,
        next_hops_fee_msat: 100,
        hop_use_fee_msat: 0,
        total_fee_msat - next_hops_fee_msat: 0,
        total_fee_msat - hop_use_fee_msat: 100,
        path_penalty_msat: 0,
        path_htlc_minimum_msat: 4294967295,
        cltv_expiry_delta: 83,
    },
    PathBuildingHop {
        node_id: NodeId(03462779ad4aad39514614751a71085f2f10e1c7a593e4e030efb5b8721ce55b0b),
        short_channel_id: 4,
        next_hops_fee_msat: 0,
        hop_use_fee_msat: 0,
        total_fee_msat - next_hops_fee_msat: 100,
        total_fee_msat - hop_use_fee_msat: 100,
        path_penalty_msat: 0,
        path_htlc_minimum_msat: 0,
        cltv_expiry_delta: 65,
    },
]
```

Matt: I'm not entirely sure what to explicitly name `total_fee_msat - [next_hops, hop_use]_fee_msat`, so I just left the variables there. Let me know if you have something else in mind.

**Comment by TheBlueMatt** [comment]
> Matt: I'm not entirely sure what to explicitly name total_fee_msat - [next_hops, hop_use]_fee_msat, so I just left the variables there. Let me know if you have something else in mind.

Oh, I meant one print of `total_fee_msat` less both the next_hops and hop_use fees. Seems strange to duplicate it.

Otherwise LGTM.

**Comment by TheBlueMatt** [comment]
Oops, sorry, no, I think we need all three - `next_hops_fee_msat`, `hop_use_fee_msat` *and* `total_fee_msat - {next_hops,hop_use]_fee_msat`.

**Review by TheBlueMatt — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]

**Pull request merged**